File: problems-extend/0501-0600/0510_inorder-successor-in-bst-ii/solution.py
class Solution:
    def inorderSuccessor(self, tree, node):
        target = None
        stack = [tree] if tree is not None else []
        while stack and target is None:
            current = stack.pop()
            if current.val == node:
                target = current
            for child in (current.left, current.right):
                if child is not None:
                    stack.append(child)
        if target is None:
            return None
        if target.right is not None:
            successor = target.right
            while successor.left is not None:
                successor = successor.left
            return successor
        while target.parent is not None and target.parent.left is not target:
            target = target.parent
        return target.parent

File: problems-extend/0501-0600/0510_inorder-successor-in-bst-ii/solution.js
/**
 * @param {NodeWithNext} tree
 * @param {number} node
 * @return {NodeWithNext}
 */
var inorderSuccessor = function (tree, node) {
    let target = null;
    const stack = [];
    if (tree !== null) stack.push(tree);
    while (stack.length > 0 && target === null) {
        const current = stack.pop();
        if (current.val === node) target = current;
        if (current.left !== null) stack.push(current.left);
        if (current.right !== null) stack.push(current.right);
    }
    if (target === null) return null;
    if (target.right !== null) {
        let successor = target.right;
        while (successor.left !== null) successor = successor.left;
        return successor;
    }
    while (target.parent !== null && target.parent.left !== target) {
        target = target.parent;
    }
    return target.parent;
};
